### PR TITLE
#4628 null check and linting fixes

### LINKED ISSUE #4628: Add null-check to tsconfig
At the moment there are a lot of possible bugs caused by incorrectly accessed objects.
Example:
```ts
let myObject: string = null;
myObject.substring(1);
```
With the current configuration there is nothing wrong with the code but this results in `TypeError: Cannot read property 'substring' of null`.

# Definition of done
- [ ] Add null-check to tsconfig `"strictNullChecks": true`.
- [ ] Refactor the bridge in order to support this change.
- [ ] Take a look at all classes and declare everything as optional that is optional.

### PR DESCRIPTION
**Pull request by Kirdock**
Closes #4628

### CODE REVIEW

File: bridge/client/app/_services/api.service.ts [hunk 1]
```diff
@@ -38,44 +39,45 @@ export class ApiService {
...
+      .get<KeptnInfoResult>(url);
   }
 
-  public getKeptnVersion(): Observable<any> {
-    let url = `${this._baseUrl}/swagger-ui/swagger.yaml`;
+  public getKeptnVersion(): Observable<string> {
+    const url = `${this._baseUrl}/swagger-ui/swagger.yaml`;
     return this.http
-      .get<any>(url, { headers: new HttpHeaders({'Access-Control-Allow-Origin': '*'}) })
+      .get<string>(url, { headers: new HttpHeaders({'Access-Control-Allow-Origin': '*'}) })
```
Comment (thisthat) [inline bridge/client/app/_services/api.service.ts:65]: I know this question is not related directly to your changes but SonarCloud is complaining about `'Access-Control-Allow-Origin': '*'`. Can we somehow restrict the `*` wildcard?

File: bridge/client/app/_components/ktb-deployment-list/ktb-deployment-list.component.html [hunk 2]
```diff
@@ -22,5 +22,5 @@
     </dt-cell>
   </ng-container>
 
-  <dt-row *dtRowDef="let row; columns: ['version', 'stages']" (click)="selectDeployment(row)" [class.active]="selectedDeployment?.shkeptncontext === row.shkeptncontext"></dt-row>
+  <dt-row *dtRowDef="let row; columns: ['version', 'stages']" (click)="selectDeployment(row)" [class.active]="selectedDeployment?.shkeptncontext === toDeployment(row).shkeptncontext"></dt-row>
```
Comment (thisthat) [inline bridge/client/app/_components/ktb-deployment-list/ktb-deployment-list.component.html:25]: Why this change?

Comment (Kirdock) [reply bridge/client/app/_components/ktb-deployment-list/ktb-deployment-list.component.html:25]: To have a type-check for `row`, because row has type `any` but is `Deployment`. If there would be a change of the property name `shkeptncontext` of `Deployment`, the IDE wouldn't find it.

File: bridge/client/app/_services/api.service.ts (hunk 1 above)
Comment (Kirdock) [reply bridge/client/app/_services/api.service.ts:65]: A just saw we can actually remove the whole `getKeptnVersion` function because this information is already in the fetched metadata.

File: bridge/client/app/_components/ktb-secrets-list/ktb-secrets-list.component.html [hunk 3]
```diff
@@ -16,7 +16,7 @@ <h2>Secrets</h2>
       <dt-table [dataSource]="tableEntries">
         <ng-container dtColumnDef="secretname" dtColumnAlign="text" dtColumnProportion="8">
           <dt-header-cell *dtHeaderCellDef>Name</dt-header-cell>
-          <dt-cell *dtCellDef="let row" [textContent]="row.name || 'n/a'"></dt-cell>
+          <dt-cell *dtCellDef="let row" [textContent]="toSecret(row).name || 'n/a'"></dt-cell>
```
Comment (mowies) [inline bridge/client/app/_components/ktb-secrets-list/ktb-secrets-list.component.html:19]: why are those "typecast" methods even needed? there are no types in the html anyways right?

File: bridge/client/app/_components/ktb-services-list/ktb-services-list.component.html [hunk 4]
```diff
@@ -4,47 +4,53 @@
...
-            <ng-template #serviceContent>
-              <span [textContent]="row.getShortImageName() || row.serviceName" [dtOverlay]="overlay" [dtOverlayConfig]="{data: {text: row.getShortImageName() || row.serviceName}}"></span>
-              <ng-container *ngIf="row.getImageVersion()">
-                <br>
-                <span [textContent]="row.getImageVersion()" [dtOverlay]="overlay" [dtOverlayConfig]="{data: {text: row.getImageVersion()}}"></span>
-              </ng-container>
-            </ng-template>
-            <ng-template #overlay let-data><span [textContent]="data.text"></span></ng-template>
-          </p>
+          <ng-container *ngIf="toService(row) as service">
```
Comment (mowies) [inline bridge/client/app/_components/ktb-services-list/ktb-services-list.component.html:7]: again here, is this really needed?

**Review by mowies — CHANGES_REQUESTED** [review]

File: bridge/client/app/_components/ktb-secrets-list/ktb-secrets-list.component.html (hunk 3 above)
Comment (Kirdock) [reply bridge/client/app/_components/ktb-secrets-list/ktb-secrets-list.component.html:19]: We have types in HTML, but the row of a DT-Table is default `any`. That's why we adjust the type.

File: bridge/client/app/_components/ktb-deployment-stage-timeline/ktb-deployment-stage-timeline.component.html [hunk 5]
```diff
@@ -1,20 +1,20 @@
-<div fxLayout="row" fxLayoutAlign="space-between" fxLayoutGap="5px" class="timeline">
-  <ng-container *ngFor="let stage of deployment.sequence.getStages()">
+<div fxLayout="row" fxLayoutAlign="space-between" fxLayoutGap="5px" class="timeline" *ngIf="deployment && deployment.sequence as sequence">
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_components/ktb-deployment-stage-timeline/ktb-deployment-stage-timeline.component.html:1]: Yes and no. It is possible but Typescript is complaining that `deployment` can be `undefined` when compiling it if we just use `deployment?.sequence`

File: bridge/client/app/_models/project.ts [hunk 6]
```diff
@@ -4,26 +4,24 @@ import {Service} from './service';
...
-  gitToken: string;
-  shipyardVersion: string;
-  allSequencesLoaded: boolean;
-
-  stages: Stage[];
-  services: Service[];
+  projectName!: string;
+  gitUser?: string;
+  gitRemoteURI?: string;
+  shipyardVersion?: string; // TODO: can this be undefined?
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_models/project.ts]: Is removed now

**Review by ermin-muratovic — CHANGES_REQUESTED** [review]

File: bridge/client/app/_components/ktb-copy-to-clipboard/ktb-copy-to-clipboard.component.ts [hunk 7]
```diff
@@ -21,10 +21,10 @@ import {Platform} from '@angular/cdk/platform';
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class KtbCopyToClipboardComponent {
-  @Input() public value: string;
-  @Input() public label: string;
+  @Input() public value?: string;
+  @Input() public label?: string;
 
-  public visible: boolean = false;
+  public visible = false;
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_components/ktb-copy-to-clipboard/ktb-copy-to-clipboard.component.ts:27]: Because we have a TSLint rule `no-inferrable-types`

File: bridge/client/app/_components/ktb-create-secret-form/ktb-create-secret-form.component.ts [hunk 8]
```diff
@@ -1,21 +1,19 @@
 import {Component, OnInit} from '@angular/core';
 import {DataService} from '../../_services/data.service';
 import {ActivatedRoute, Router} from '@angular/router';
-import {FormControl, FormGroup, Validators} from "@angular/forms";
-import {Secret} from "../../_models/secret";
+import { AbstractControl, FormControl, FormGroup, Validators } from '@angular/forms';
+import {Secret} from '../../_models/secret';
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_components/ktb-create-secret-form/ktb-create-secret-form.component.ts:5]: The linter does not complain in both cases

File: bridge/client/app/_components/ktb-evaluation-details/ktb-evaluation-details.component.ts [hunk 9]
```diff
@@ -15,69 +33,58 @@ noData(Highcharts);
...
-  public _evaluationData: Trace;
-  public _selectedEvaluationData: Trace;
-  public _comparisonView: string = 'heatmap';
-  private _metrics: string[];
-
+  public _evaluationState: Map<ResultTypes, string> = new Map<ResultTypes, string>([
+    [ResultTypes.PASSED, 'recovered'],
+    [ResultTypes.WARNING, 'warning'],
+    [ResultTypes.FAILED, 'error'],
+    // ['failed', 'error'] TODO: Why is this here? Was this different some time ago?
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_components/ktb-evaluation-details/ktb-evaluation-details.component.ts]: done

File: bridge/client/app/_components/ktb-notification-bar/ktb-notification-bar.component.ts [hunk 10]
```diff
@@ -1,24 +1,22 @@
-import { Component, OnInit } from '@angular/core';
-import {Observable} from "rxjs";
-import {Notification, TemplateRenderedNotifications} from "../../_models/notification";
-import {NotificationsService} from "../../_services/notifications.service";
+import { Component } from '@angular/core';
+import {Observable} from 'rxjs';
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_components/ktb-notification-bar/ktb-notification-bar.component.ts:2]: The one with the spaces yes.
rule:
```json
"whitespace": {
      "options": [
        "check-module"
      ]
    }
````
Should I add it?

[the remark this answers is not in the document]
Comment (mowies) [reply bridge/client/app/_components/ktb-notification-bar/ktb-notification-bar.component.ts:2]: I would vote for adding it. Although this will be handled by prettier in the future.

File: bridge/client/app/_components/ktb-stage-overview/ktb-stage-overview.component.ts [hunk 11]
```diff
@@ -114,9 +122,12 @@ export class KtbStageOverviewComponent implements OnInit, OnDestroy {
     this._changeDetectorRef.markForCheck();
   }
 
-  public filterChanged(event: any) {
+  // tslint:disable-next-line:no-any
+  public filterChanged(event: DtFilterFieldChangeEvent<any>) {
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_components/ktb-stage-overview/ktb-stage-overview.component.ts:126]: I can't define a custom type because then I receive an error `Argument type ... is not assignable to parameter type ...`, but I'll introduce some casting and new types.

File: bridge/client/app/_components/ktb-notification-bar/ktb-notification-bar.component.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ermin-muratovic) [reply bridge/client/app/_components/ktb-notification-bar/ktb-notification-bar.component.ts:2]: Yep, let's add this, as we also agreed on using the syntax with spaces.

File: bridge/client/app/_utils/test.utils.ts [hunk 12]
```diff
@@ -1,6 +1,6 @@
 export class TestUtils {
   public static createNewDropEventWithFiles(files: File[]): DragEvent {
-    const dataTransfer: DataTransfer = new MockDataTransfer(files);
+    const dataTransfer: DataTransfer = MockDataTransfer(files);
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_utils/test.utils.ts:3]: In my opinion we can leave it there because the file name is `test-utils.ts` and already indicates that it is for testing purposes

[the remark this answers is not in the document]
Comment (mowies) [reply bridge/client/app/_utils/test.utils.ts:3]: if the file is ignored during compilation I am fine with that

File: bridge/client/app/_components/ktb-keptn-services-list/ktb-keptn-services-list.component.ts [hunk 13]
```diff
@@ -53,8 +53,12 @@ export class KtbKeptnServicesListComponent implements OnInit, OnDestroy {
       })
     ).subscribe((uniformRegLogs) => {
       uniformRegLogs.sort((a, b) => {
-        if (a.time.valueOf() > b.time.valueOf()) return -1;
-        if (a.time.valueOf() < b.time.valueOf()) return 1;
+        if (a.time.valueOf() > b.time.valueOf()) {
```
[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_components/ktb-keptn-services-list/ktb-keptn-services-list.component.ts]: done

File: bridge/client/app/_utils/test.utils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (laneli) [reply bridge/client/app/_utils/test.utils.ts:3]: `tsconfig.app.json` has a dedicated exclude property for that:
https://www.typescriptlang.org/tsconfig#exclude

[the remark this answers is not in the document]
Comment (Kirdock) [reply bridge/client/app/_utils/test.utils.ts:3]: done

File: bridge/tsconfig.app.json [hunk 14]
```diff
@@ -13,6 +13,7 @@
   ],
   "exclude": [
     "client/test.ts",
-    "client/**/*.spec.ts"
+    "client/**/*.spec.ts",
+    "client/app/_utils/test.utils.ts"
```
Comment (mowies) [inline bridge/tsconfig.app.json:17]: Maybe we should also add `client/**/*.mock.ts` here

Comment (Kirdock) [reply bridge/tsconfig.app.json:18]: done

**Review by ermin-muratovic — DISMISSED** [review]
:)

**Review by mowies — APPROVED** [review]
Good job! LGTM

File: bridge/client/app/_pipes/to-type.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (mowies) [inline bridge/client/app/_pipes/to-type.ts:1]: so much cleaner with a pipe than the cast-methods, love it 👌

**Review by mowies — APPROVED** [review]

**Review by mowies — APPROVED** [review]

File: bridge/client/app/_pipes/to-type.ts (hunk 15 above)
Comment (ermin-muratovic) [reply bridge/client/app/_pipes/to-type.ts:1]: Yep, that's much better :)
